### ISSUE TITLE
VLA-JEPA: hard-coded last-layer cond tap reads untrained token positions when freeze_qwen=True

### ISSUE THREAD

**Issue by OliverAltergott** [huggingface/lerobot#4414]
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
Not runtime-environment specific — a design/wiring issue in the VLA-JEPA cond path, present on
main @ 59ab28620f3f. Found and measured on lerobot installed from
git+https://github.com/huggingface/lerobot@2d7a42011a4f8e05a8c85d5fb908da258d4cc7b1.
Base checkpoint lerobot/VLA-JEPA-Pretrain (Qwen/Qwen3-VL-2B-Instruct, 28 language layers),
fine-tuned with freeze_qwen=True, enable_world_model=False, onto a 2-camera SO-101 tabletop arm.
```

### Description

`freeze_qwen=True` is a supported configuration, but in that mode the action head's conditioning is read from token positions whose meaning was fixed by the *source* checkpoint's training domain and can no longer adapt — and the tap is hard-coded, so there is no way to point it somewhere better.

Two facts combine badly:

1. **The tap is hard-coded to the last decoder layer, at the `<|embodied_action|>` token positions.** `VLAJEPAModel._qwen_last_decoder_hidden` (`src/lerobot/policies/vla_jepa/modeling_vla_jepa.py:154-169`) hooks `self.qwen.model.model.language_model.layers[-1]`, and `predict_action`/`forward` then gather only the embodied-token positions out of it (`modeling_vla_jepa.py:206`). Neither the layer index nor the position set is exposed in `VLAJEPAConfig`.

2. **Those positions are learned readouts, and `freeze_qwen=True` freezes what they learned somewhere else.** `<|embodied_action|>` is added by `Qwen3VLInterface.expand_tokenizer` (`qwen_interface.py:68-74`); it has no meaning in stock Qwen3-VL and acquires one only while the backbone trains. In a released checkpoint such as `lerobot/VLA-JEPA-Pretrain` or `lerobot/VLA-JEPA-LIBERO` (both `freeze_qwen: False`, `enable_world_model: True`) it has acquired a good one — for *that* checkpoint's robot, cameras and scene. `config.freeze_qwen` then does `self.qwen.requires_grad_(False)` (`modeling_vla_jepa.py:134`), so on a new embodiment those readouts cannot re-adapt. They are not random, and they are not empty; they are tuned for a domain the user is no longer in, and pinned there.

The design is sound with the default `freeze_qwen=False` (`configuration_vla_jepa.py:44`): the embodied tokens learn, through all 28 layers, to gather whatever the head needs, and the last layer is the natural place to collect the result. Frozen, they are inherited query slots read at the most abstract point in the network — precisely where a language model has already traded spatial detail for semantics, and precisely where a domain shift bites hardest.

Image-token positions a few layers in behave the opposite way: they carry spatial detail that transfers without any adaptation at all. Nothing in the config lets a user find that out or act on it.

### Context & Reproduction

The hard-coded tap, `modeling_vla_jepa.py:154-169` and `:206`:

```python
    def _qwen_last_decoder_hidden(self, qwen_inputs: dict[str, torch.Tensor]) -> torch.Tensor:
        ...
        last_layer = self.qwen.model.model.language_model.layers[-1]
        handle = last_layer.register_forward_hook(_hook)
        ...
            embodied_action_tokens = last_hidden[embodied_idx[0], embodied_idx[1], :].view(b, -1, h)
```

**Measured, same frozen backbone, one forward pass with layer hooks at several depths and both position sets.** Task: recover the shoulder-pan angle the grasp requires from a single frame (a 6-DoF arm reaching for a ball on a table), grouped leave-one-episode-out ridge over 199 episodes. Chance is 39.6°; a 1-NN readout on the ball's own pixels gets 10.7°, which is the practical ceiling for this scene.

| tap | error | notes |
|---|---|---|
| embodied tokens, layer 27 (**as shipped**) | 28.1° | best of the embodied-position depths — 3/7/11/15/19/23 are all flat or worse |
| image tokens, layer 3 (mean-pooled 5×5/cam) | **21.8°** | spatial detail decays monotonically with depth at image positions: L3 21.8°, L15 25.1°, L27 31.8° |
| vision-tower output | 37.2° | the first few LM layers do something useful; the last 24 undo it |

Positive control (predicting the arm's *own* current pan) is strong at every tap, so no depth's result is a power artefact. Concatenating taps does not help; a finer pool (5×10) does not either.

Retraining the action head on the layer-3 image-token tap, changing nothing else — same base checkpoint, same data, same schedule, same proprio-dropout, same seed — moved held-out aim from **29% to 71%** of the commanded motion the grasp needs, against an otherwise identical layer-27 control. It also erased a train-vs-held-out gap we had spent weeks attributing to overfitting.

Worth noting on the base checkpoint: this is `VLA-JEPA-Pretrain`, not `VLA-JEPA-LIBERO` — real-robot exterior cameras rather than simulation, i.e. the closest released starting point to our setup and the one whose embodied readouts should transfer best. The gap above is what remains *after* choosing well. A frozen readout is still a readout tuned on someone else's robot, and today the user has no way to measure that or route around it.

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

Suggested fixes, cheapest first:

1. **Warn when `freeze_qwen=True`.** `VLAJEPAConfig.__post_init__` already special-cases `freeze_qwen and enable_world_model` (`configuration_vla_jepa.py:105-107`). One more branch, noting that the cond tap reads frozen readouts inherited from the source checkpoint's domain, would have saved us the whole investigation.
2. **Expose the tap in the config** — a layer index and a choice of position set (`embodied` | `image`), defaulting to the current `(-1, embodied)` so nothing changes for existing users. This is a small change; the hook in `_qwen_last_decoder_hidden` already generalises to any `layers[i]`, and the image positions are recoverable from the processor's image-token id and grid.
3. Optionally document that `freeze_qwen=True` is a cost optimisation with a real accuracy cost in the shipped configuration, rather than a free one.

Happy to open a PR for (1) and (2) if that's a direction you'd take.

**Comment by OliverAltergott** [comment]
Some context on why `freeze_qwen=True` isn't a corner case, which I under-sold in the original report.

We freeze the backbone deliberately, not to save compute. The reason to start from a released VLA-JEPA checkpoint at all is everything already in it — Qwen3-VL's visual-language grounding, plus whatever the end-to-end pretraining put there — and fine-tuning a 2B backbone on ~110k frames of a single tabletop workspace is a reliable way to lose it. You get a model that is better at your three object positions and worse at everything else. LoRA is the usual middle ground, but it is still a modification of the backbone, and `VLAJEPAConfig` doesn't expose one in any case. So "take the released checkpoint as-is, train only the action head" is the configuration we actively want, and `freeze_qwen=True` is the flag that expresses it. (For completeness: our own runs set `enable_world_model=False`, so we train neither the V-JEPA2 encoder nor the action-conditioned predictor — the action head is the only thing being trained. The base checkpoint was trained with the world model on, so its effect is inherited through the frozen weights.)

That is what makes the hard-coded tap awkward rather than merely inflexible. In the frozen regime the only trainable path from pixels to actions runs through the action head, so *where the head reads from* becomes the single most consequential wiring choice in the policy — and it is the one thing a user cannot change. The shipped default is the right choice for a regime (trainable backbone, embodied tokens free to re-learn what the new task needs) that a user setting `freeze_qwen=True` has explicitly opted out of.

Two reasonable resolutions, and I'd be happy with either:

- If VLA-JEPA is intended to be trained end-to-end and `freeze_qwen` exists mainly for debugging, that is a legitimate design position — but it's worth one line in the config docstring or the policy README. It would have saved us the investigation.
- If frozen-backbone use is meant to be supported, then `(layer, positions)` should be a config field, defaulting to today's `(-1, embodied)` so nothing changes for anyone.

**Comment by OliverAltergott** [comment]
One more argument for the warning in suggestion (1), which I think makes it a smaller ask than I framed it as: the codebase already contains exactly this reasoning, applied to a different component.

`VLAJEPAConfig.__post_init__` (`configuration_vla_jepa.py:105-107`):

```python
if self.freeze_qwen and self.enable_world_model:
    # freezing qwen backbone makes world model training irrelevant since no grad flows
    self.enable_world_model = False
```

That is precisely the right call, and the comment states the general principle: freezing the backbone severs the gradient path, so any component that depends on that path stops doing its job. The world model is handled. The cond tap has the same dependency and is not.

The `<|embodied_action|>` positions are learned readouts — they mean nothing in stock Qwen3-VL and acquire their meaning only while the backbone trains. In a released checkpoint they carry a good one, for that checkpoint's robot and cameras. `freeze_qwen=True` pins them there, and unlike the world model there is no branch that notices: the tap keeps reading positions tuned for the source domain, and the run trains to completion with nothing in the config, the logs, or the metrics indicating anything is off. Our first symptom was a train/held-out gap that looked exactly like overfitting, which we chased for some time before the tap turned out to be the cause.

So the change I'd propose for (1) is a sibling of the branch above, in the same method:

```python
if self.freeze_qwen:
    logging.warning(
        "freeze_qwen=True: the action-head conditioning is read from the <|embodied_action|> "
        "positions at the last decoder layer. Those are learned readouts, fixed by whatever "
        "the source checkpoint was trained on, and they cannot adapt to your embodiment while "
        "the backbone is frozen. Expect reduced conditioning quality under domain shift."
    )
```

Same method, same flag, same principle — just extended to the second component it applies to. Suggestion (2), making the tap configurable, is what actually lets a user act on that warning, but the warning alone would have redirected our investigation by weeks.
